subsys/bluetooth/scan.c

```diff
@@ -283,6 +306,71 @@ static void notify_connecting_error(struct bt_scan_device_info *device_info)
 	}
 }
 
+static void attempts_filter_force_add(struct conn_attempts_filter *filter,
+				      const bt_addr_le_t *addr)
+{
+	LOG_DBG("Force adding new device filter");
+
+	/* Overwrite the oldest device */
+	memset(filter->oldest_device, 0, sizeof(*filter->oldest_device));
+	memcpy(&filter->oldest_device->addr, addr, sizeof(*addr));
+
+	if (filter->oldest_device == &filter->device[filter->count - 1]) {
+		filter->oldest_device = filter->device;
+
+		return;
+	}
+
+	filter->oldest_device++;
+}
+
+static void device_conn_attempts_count(struct bt_conn *conn)
+{
+	const bt_addr_le_t *addr = bt_conn_get_dst(conn);
+	struct conn_attempts_filter *filter = &bt_scan.attempts_filter;
+
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+
+	for (size_t i = 0; i < filter->count; i++) {
+		struct conn_attempts_device *device = &filter->device[i];
+
+		if (bt_addr_le_cmp(addr, &device->addr) == 0) {
+			if (device->attempts < CONFIG_BT_SCAN_CONNECTION_ATTEMPTS_CNT) {
+				device->attempts++;
+
+				break;
+			}
+		}
+	}
+
+	k_mutex_unlock(&conn_attempts_mutex);
+}
+
+static bool conn_attempts_filter_check(const bt_addr_le_t *addr)
+{
+	struct conn_attempts_filter *filter = &bt_scan.attempts_filter;
+	bool attempts_exceeded = false;
+
+	k_mutex_lock(&conn_attempts_mutex, K_FOREVER);
+
+	/* Check if the device is in the filter array. */
+	for (size_t i = 0; i < filter->count; i++) {
+		struct conn_attempts_device *device = &filter->device[i];
+
+		if (bt_addr_le_cmp(addr, &device->addr) == 0) {
+			if (device->attempts >= CONFIG_BT_SCAN_CONNECTION_ATTEMPTS_CNT) {
+				attempts_exceeded = true;
+			}
+
+			break;
+		}
+	}
+
+	k_mutex_unlock(&conn_attempts_mutex);
+
+	return !attempts_exceeded;
+}
+
 static void scan_connect_with_target(struct bt_scan_control *control,
```

Comment:
Should we notify that filters matched for the blacklisted device?

Reply:
I think we should not notify about any event when connection attempts is exceeded or device is on the blacklist